File: frontend/src/utils/mcp-config.ts
import {
  MCPConfig,
  MCPSSEServer,
  MCPSHTTPServer,
  MCPStdioServer,
  SettingsValue,
} from "#/types/settings";

const EMPTY_MCP_CONFIG: MCPConfig = {
  sse_servers: [],
  stdio_servers: [],
  shttp_servers: [],
};

type SdkMcpServerConfig = Record<string, SettingsValue>;
type SdkMcpConfig = { mcpServers: Record<string, SdkMcpServerConfig> };

/**
 * Parse an SDK mcp_config value ({ mcpServers: { ... } }) and convert it
 * to the frontend MCPConfig format used by UI components.
 */
export function parseMcpConfig(value: unknown): MCPConfig {
  if (!value || typeof value !== "object") {
    return { ...EMPTY_MCP_CONFIG };
  }

  const obj = value as Record<string, unknown>;

  if (
    !("mcpServers" in obj) ||
    !obj.mcpServers ||
    typeof obj.mcpServers !== "object"
  ) {
    return { ...EMPTY_MCP_CONFIG };
  }

  const sseServers: (string | MCPSSEServer)[] = [];
  const stdioServers: MCPStdioServer[] = [];
  const shttpServers: (string | MCPSHTTPServer)[] = [];

  const mcpServers = obj.mcpServers as Record<string, Record<string, unknown>>;

  for (const [serverName, serverConfig] of Object.entries(mcpServers)) {
    // eslint-disable-next-line no-continue
    if (!serverConfig || typeof serverConfig !== "object") continue;

    const url = serverConfig.url as string | undefined;

    if (url) {
      const transport = serverConfig.transport as string | undefined;
      const auth = serverConfig.auth as string | undefined;
      const apiKey =
        typeof auth === "string" && auth !== "oauth" ? auth : undefined;

      if (transport === "sse") {
        const server: MCPSSEServer = { url };
        if (apiKey) server.api_key = apiKey;
        sseServers.push(server);
      } else {
        const server: MCPSHTTPServer = { url };
        if (apiKey) server.api_key = apiKey;
        if (serverConfig.timeout != null) {
          server.timeout = serverConfig.timeout as number;
        }
        shttpServers.push(server);
      }
    } else {
      const stdioServer: MCPStdioServer = {
        name: serverName,
        command: serverConfig.command as string,
      };
      if (serverConfig.args) {
        stdioServer.args = serverConfig.args as string[];
      }
      if (serverConfig.env) {
        stdioServer.env = serverConfig.env as Record<string, string>;
      }
      stdioServers.push(stdioServer);
    }
  }

  return {
    sse_servers: sseServers,
    stdio_servers: stdioServers,
    shttp_servers: shttpServers,
  };
}

/**
 * Convert the frontend MCPConfig format back to the SDK { mcpServers: { ... } }
 * shape expected by agent_settings.mcp_config on the backend.
 */
export function toSdkMcpConfig(config: MCPConfig): SdkMcpConfig | null {
  const mcpServers: Record<string, SdkMcpServerConfig> = {};
  let counter = 0;

  const nextName = (base: string): string => {
    const name = counter === 0 ? base : `${base}_${counter}`;
    counter += 1;
    return name;
  };

  for (const entry of config.sse_servers) {
    const server: SdkMcpServerConfig = {};
    if (typeof entry === "string") {
      server.url = entry;
    } else {
      server.url = entry.url;
      if (entry.api_key) server.auth = entry.api_key;
    }
    server.transport = "sse";
    mcpServers[nextName("sse")] = server;
  }

  for (const entry of config.shttp_servers) {
    const server: SdkMcpServerConfig = {};
    if (typeof entry === "string") {
      server.url = entry;
    } else {
      server.url = entry.url;
      if (entry.api_key) server.auth = entry.api_key;
      if (entry.timeout != null) server.timeout = entry.timeout;
    }
    mcpServers[nextName("shttp")] = server;
  }

  for (const entry of config.stdio_servers) {
    const server: SdkMcpServerConfig = {
      command: entry.command,
    };
    if (entry.args) server.args = entry.args;
    if (entry.env) server.env = entry.env;
    mcpServers[nextName(entry.name || "stdio")] = server;
  }

  return Object.keys(mcpServers).length > 0 ? { mcpServers } : null;
}

File: openhands/core/config/mcp_config.py
"""MCP configuration — thin wrappers around the SDK MCPConfig from *fastmcp*.

All server configuration uses the unified ``MCPConfig.mcpServers`` dict.
Legacy helpers (``from_toml_section``, ``merge``) are provided for
config.toml parsing and server merging.
"""

from __future__ import annotations

import os
import shlex
from typing import TYPE_CHECKING, Any

from fastmcp.mcp_config import MCPConfig, RemoteMCPServer, StdioMCPServer

if TYPE_CHECKING:
    from openhands.core.config.openhands_config import OpenHandsConfig

from openhands.core.logger import openhands_logger as logger
from openhands.utils.import_utils import get_impl

__all__ = [
    'MCPConfig',
    'RemoteMCPServer',
    'StdioMCPServer',
    'OpenHandsMCPConfig',
    'OpenHandsMCPConfigImpl',
    'merge_mcp_configs',
    'mcp_config_from_toml',
]


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def merge_mcp_configs(base: MCPConfig, other: MCPConfig) -> MCPConfig:
    """Return a new MCPConfig with servers from both configs merged."""
    merged = dict(base.mcpServers)
    merged.update(other.mcpServers)
    return MCPConfig(mcpServers=merged)


def _parse_stdio_args(v: Any) -> list[str]:
    """Parse stdio args from a string using shlex or return list as-is."""
    if isinstance(v, str):
        if not v.strip():
            return []
        return shlex.split(v.strip())
    return list(v or [])


def _parse_stdio_env(v: Any) -> dict[str, str]:
    """Parse stdio env from a comma-separated string or return dict as-is."""
    if isinstance(v, str):
        env: dict[str, str] = {}
        for pair in v.split(','):
            pair = pair.strip()
            if not pair:
                continue
            if '=' not in pair:
                raise ValueError(
                    f"Environment variable '{pair}' must be in KEY=VALUE format"
                )
            key, value = pair.split('=', 1)
            env[key.strip()] = value
        return env
    return dict(v or {})


def mcp_config_from_toml(data: dict[str, Any]) -> dict[str, MCPConfig]:
    """Parse a ``[mcp]`` TOML section into ``{'mcp': MCPConfig}``.

    Accepts the legacy ``sse_servers`` / ``shttp_servers`` / ``stdio_servers``
    list format and converts to the unified ``mcpServers`` dict.
    """
    servers: dict[str, RemoteMCPServer | StdioMCPServer] = {}

    for entry in data.get('sse_servers', []):
        if isinstance(entry, str):
            entry = {'url': entry}
        name = f'sse_{len([k for k in servers if k.startswith("sse_")])}'
        servers[name] = RemoteMCPServer(
            url=entry['url'],
            transport='sse',
            auth=entry.get('api_key'),
        )

    for entry in data.get('shttp_servers', []):
        if isinstance(entry, str):
            entry = {'url': entry}
        name = f'shttp_{len([k for k in servers if k.startswith("shttp_")])}'
        servers[name] = RemoteMCPServer(
            url=entry['url'],
            transport='http',
            auth=entry.get('api_key'),
            timeout=entry.get('timeout', 60),
        )

    for entry in data.get('stdio_servers', []):
        name = entry.get(
            'name', f'stdio_{len([k for k in servers if k.startswith("stdio_")])}'
        )
        servers[name] = StdioMCPServer(
            command=entry['command'],
            args=_parse_stdio_args(entry.get('args', [])),
            env=_parse_stdio_env(entry.get('env', {})),
        )

    return {'mcp': MCPConfig(mcpServers=servers)}


# ---------------------------------------------------------------------------
# OpenHands default MCP server factory
# ---------------------------------------------------------------------------


class OpenHandsMCPConfig:
    """Factory for the default OpenHands MCP server entries."""

    @staticmethod
    def add_search_engine(
        app_config: 'OpenHandsConfig',
    ) -> dict[str, StdioMCPServer] | None:
        """Return a tavily stdio server entry if a Tavily API key is configured."""
        if (
            app_config.search_api_key
            and app_config.search_api_key.get_secret_value().startswith('tvly-')
        ):
            logger.info('Adding search engine to MCP config')
            return {
                'tavily': StdioMCPServer(
                    command='npx',
                    args=['-y', 'tavily-mcp@0.2.1'],
                    env={
                        'TAVILY_API_KEY': app_config.search_api_key.get_secret_value()
                    },
                )
            }
        logger.warning('No search engine API key found, skipping search engine')
        return None

    @staticmethod
    async def create_default_mcp_server_config(
        host: str, config: 'OpenHandsConfig', user_id: str | None = None
    ) -> dict[str, RemoteMCPServer | StdioMCPServer]:
        """Return a dict of default MCP server entries to merge into config.mcp.

        Returns:
            dict mapping server names to their configs.
        """
        servers: dict[str, RemoteMCPServer | StdioMCPServer] = {}

        search = OpenHandsMCPConfig.add_search_engine(config)
        if search:
            servers.update(search)

        servers['openhands'] = RemoteMCPServer(
            url=f'http://{host}/mcp/mcp',
            transport='http',
            timeout=60,
        )

        return servers


openhands_mcp_config_cls = os.environ.get(
    'OPENHANDS_MCP_CONFIG_CLS',
    'openhands.core.config.mcp_config.OpenHandsMCPConfig',
)

OpenHandsMCPConfigImpl = get_impl(OpenHandsMCPConfig, openhands_mcp_config_cls)
